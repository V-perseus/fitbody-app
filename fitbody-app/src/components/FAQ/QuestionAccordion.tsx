import React, { useState } from 'react'
import { Animated, TouchableOpacity, Text, View } from 'react-native'
import LinearGradient from 'react-native-linear-gradient'

// Assets
import styles from './styles'
import ChevronDown from '../../../assets/images/svg/icon/16px/cheveron/down.svg'
import ChevronUp from '../../../assets/images/svg/icon/16px/cheveron/up.svg'
import globals from '../../config/globals'

interface IQuestionAccordionProps {
  title: string
  children: React.ReactNode
  slideLayoutAnimation: () => void
}
const QuestionAccordion: React.FC<IQuestionAccordionProps> = ({ title, children, slideLayoutAnimation }) => {
  const [extended, setExtended] = useState(false)

  /**
   * Handle Click on question section
   */
  function handleOpenCloseState() {
    slideLayoutAnimation()
    setExtended(!extended)
  }

  return (
    <TouchableOpacity onPress={handleOpenCloseState} style={styles.question} testID="faq_btn">
      {/* Outer expandable container */}
      <View>
        {/* question */}
        <View style={styles.question}>
          <LinearGradient
            style={extended ? styles.questionHeaderActive : styles.questionHeader}
            colors={extended ? [globals.styles.colors.colorPink, globals.styles.colors.colorLove] : ['transparent', 'transparent']}>
            <Text
              numberOfLines={extended ? 200 : 2}
              ellipsizeMode="tail"
              style={extended ? styles.textFieldsQuestionActive : styles.textFieldsQuestion}>
              {title}
            </Text>
            <View style={styles.arrow}>
              {extended ? (
                <ChevronUp color={globals.styles.colors.colorWhite} />
              ) : (
                <ChevronDown color={globals.styles.colors.colorGrayDark} />
              )}
            </View>
          </LinearGradient>
        </View>
        {/* Animated answer section */}
        <Animated.View>
          <View>
            {extended ? (
              <TouchableOpacity>
                <View style={styles.answerBody}>{children}</View>
              </TouchableOpacity>
            ) : null}
          </View>
        </Animated.View>
      </View>
    </TouchableOpacity>
  )
}

export default QuestionAccordion
