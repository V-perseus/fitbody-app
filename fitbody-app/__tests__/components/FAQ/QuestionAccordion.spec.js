import React from 'react'
import { cleanup, render, fireEvent } from '@testing-library/react-native'
import renderer from 'react-test-renderer'

import QuestionAccordion from '../../../src/components/FAQ/QuestionAccordion'

describe('FAQ QuestionAccordion', () => {
  afterEach(cleanup)

  it('renders snapshot correctly', () => {
    const tree = renderer
      .create(
        <QuestionAccordion key={`key`} title={'TITLE'} slideLayoutAnimation={() => {}}>
          BODY TEXT
        </QuestionAccordion>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render title', () => {
    const { queryByText } = render(
      <QuestionAccordion key={`key`} title={'TITLE'} slideLayoutAnimation={() => {}}>
        BODY TEXT
      </QuestionAccordion>,
    )
    expect(queryByText('TITLE')).not.toBeNull()
  })

  it('should show/hide body when pressed', async () => {
    const { queryByTestId, queryByText } = render(
      <QuestionAccordion key={`key`} title={'TITLE'} slideLayoutAnimation={() => {}}>
        BODY TEXT
      </QuestionAccordion>,
    )
    const button = queryByTestId('faq_btn')
    expect(button).toBeDefined()
    await fireEvent.press(button)
    expect(queryByText(/BODY TEXT/)).toBeDefined()
    await fireEvent.press(button)
    expect(queryByText(/BODY TEXT/)).toBeNull()
  })
})
