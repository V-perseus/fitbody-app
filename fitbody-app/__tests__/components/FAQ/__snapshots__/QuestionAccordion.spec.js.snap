// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAQ QuestionAccordion renders snapshot correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "borderBottomColor": "#dedede",
      "borderBottomWidth": 1,
      "opacity": 1,
    }
  }
  testID="faq_btn"
>
  <View>
    <View
      style={
        {
          "borderBottomColor": "#dedede",
          "borderBottomWidth": 1,
        }
      }
    >
      <BVLinearGradient
        colors={
          [
            0,
            0,
          ]
        }
        endPoint={
          {
            "x": 0.5,
            "y": 1,
          }
        }
        locations={null}
        startPoint={
          {
            "x": 0.5,
            "y": 0,
          }
        }
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 70,
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={2}
          style={
            {
              "color": "#333333",
              "flexWrap": "wrap",
              "fontFamily": "OpenSans",
              "fontSize": 14,
              "lineHeight": 20,
              "paddingBottom": 10,
              "paddingLeft": 20,
              "paddingRight": 8,
              "paddingTop": 10,
              "width": 675,
            }
          }
        >
          TITLE
        </Text>
        <View
          style={
            {
              "justifyContent": "center",
              "width": 75,
            }
          }
        >
          <down.svg
            color="#bbbbbb"
          />
        </View>
      </BVLinearGradient>
    </View>
    <View
      collapsable={false}
      style={{}}
    >
      <View />
    </View>
  </View>
</View>
`;
